### ISSUE TITLE
The process of reconstructing a 3D model using multi-view images of my face

### ISSUE THREAD

**Issue by 76donghyun** [aejion/NeuFace#13]
To reconstruct a 3D model using my facial images, I've seen that using ImFace is necessary for cropping the face mesh and rendering a mask for facial images.

Is it correct to use the "NeuFace/data_preprocess/cut_mesh.py" and "NeuFace/data_preprocess/render_mask.py" codes for the necessary steps in that process?

In cut_mesh.py code,
tu_mesh_path = '' # tu model path
high_mesh_path = '' # origin mesh path (high quality)
output_mesh_path = '' # output mesh path

What kind of file should I put in the tu model path?
Is it appropriate to put multi-view images on the original mesh path?

In render_mask.py code, 
origin_path = r'' # origin image path
mask_path = r'./mask' # output mask path
mask_image_path = r'./image' # output image path
ply_path = r'' # cut mesh path (high quality)
tu_mesh_path = '' # tu mesh path

I would appreciate it if you could let me know what files go into each of the paths mentioned above.

**Comment by aejion** [comment]
In cut_mesh.py, 'tu_mesh_path' refers to the path of the registered mesh in 'fsmview_trainset_shape_xxx.zip' of the FaceScape dataset. 'high_mesh_path' indicates the path of the high-quality mesh in 'fsmview_trainset_images_xxx.zip' of the FaceScape dataset. 'output_mesh_path' denotes the output path for the cut mesh, and you can set it to your desired location.

In render_mask.py: 'origin_path' denotes the path of the original images of the FaceScape dataset. 'ply_path' refers to the 'output_mesh_path' from cut_mesh.py. 'tu_mesh_path' remains the same as in cut_mesh.py.

**Comment by 76donghyun** [comment]
When using my own facial images, where should I place them? Which files from the FaceScape dataset should be replaced, and which ones should be used as they are? Does the reconstruction of my face in 3D necessarily require some parts of the FaceScape dataset?

**Comment by aejion** [comment]
When using your own images, I recommend obtaining the training dataset through the following steps: 1) using your multi-view images to reconstruct a mesh registered to the FaceScape bilinear model. 2) Apply the ‘data_preprocess’ code to crop the mesh obtained in the first step. 3) Use the cropped mesh to do masking operations on the original images. Finally, you can get the training dataset.

**Comment by 76donghyun** [comment]
Is the output of the FaceScape bilinear model the high_mesh(origin_mesh)? If so, how is tu_mesh generated? Is it created using my own facial images, or does it rely on the FaceScape dataset?

You mentioned creating a training dataset. Can I use my facial images to test the pre-trained model through a similar process? Is it not possible to perform testing on my facial image using the pre-trained model?

**Comment by aejion** [comment]
Firstly, NeuFace is a reconstruction method rather than a generative one. Therefore, for each dataset, it requires separate training. The provided pre-trained model corresponds to the results obtained from training on a specific identity of FaceScape Dataset. 

Secondly, during training, if you are using the FaceScape dataset, it comes with high-quality meshes and TU meshes. You can directly utilize the preprocessing code to process and obtain the training dataset. If you want to use your own dataset for training, you need to obtain the corresponding mesh for your dataset, then register the mesh to the FaceScape space. Once you have the aligned mesh, you can refer to the preprocessing code to crop the registered mesh. Then, using this cropped mesh to create masks for your multi-view images.

**Comment by 76donghyun** [comment]
In conclusion, are you saying that to perform 3D neural face rendering using my images, I need the 3D mesh of my face? I understood that obtaining such a mesh is done through the FaceScape bilinear model, is that correct?

How can I obtain the TU mesh for my facial image?

**Comment by aejion** [comment]
Yes, you need the 3D mesh to mask the original image. If you use the FaceScape bilinear model to fit your mesh, the output of the bilinear model is tu mesh.

**Comment by 76donghyun** [comment]
I sincerely express my gratitude for your kind response. What are the processes and codes to generate a 3D mesh and TU mesh, respectively, from my facial image, as mentioned below?

[bilinear_model-basic](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_bilinear_basic.ipynb) - use facescape bilinear model to generate 3D mesh models.
[bilinear_model-fit](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_bilinear_fit.ipynb) - fit the bilinear model to 2D/3D landmarks.
[multi-view-project](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_mview_projection.ipynb) - Project 3D models to multi-view images.
[landmark](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_landmark.ipynb) - extract landmarks using predefined vertex index.
[facial_mask](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_mask.ipynb) - extract facial region from the full head TU-models.
[render](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_render.ipynb) - render TU-models to color images and depth map.
[alignment](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_align.ipynb) - align all the multi-view models.
[symmetry](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_symmetry.ipynb) - get the correspondence of the vertices on TU-models from left side to [rig](https://nbviewer.jupyter.org/github/zhuhao-nju/facescape/blob/master/toolkit/demo_rig.ipynb)ht side.
rig - rig 20 expressions to 52 expressions.
